--- a/debian/rules	2009-09-04 16:11:28.000000000 +0900
+++ b/debian/rules	2009-09-04 16:12:12.000000000 +0900
@@ -22,11 +22,16 @@
 
 # implement no optimization build option
 ifneq (,$(findstring noopt,$(DEB_BUILD_OPTIONS)))
-CFLAGS = "-g -DSYSLOG_SUCCESS -DSYSLOG_FAILURE -DSYSLOG_NON_ROOT -O0"
+CFLAGS = -g -DSYSLOG_SUCCESS -DSYSLOG_FAILURE -DSYSLOG_NON_ROOT -O0
 else
-CFLAGS = "-g -DSYSLOG_SUCCESS -DSYSLOG_FAILURE -DSYSLOG_NON_ROOT -O2"
+CFLAGS = -g -DSYSLOG_SUCCESS -DSYSLOG_FAILURE -DSYSLOG_NON_ROOT -O2
+endif
+LDFLAGS =
+
+# Renesas SH(sh4) need -mieee option.
+ifeq ($(DEB_HOST_ARCH_CPU),sh4)
+  CFLAGS += -mieee
 endif
-LDFLAGS = 
 
 # programs to include/not include
 ifeq ($(DEB_HOST_ARCH_OS),hurd)
