--- coreutils-4.5.3.orig/src/echo.c
+++ coreutils-4.5.3/src/echo.c
@@ -81,13 +81,12 @@
   -n              do not output the trailing newline\n\
   -e              enable interpretation of the backslash-escaped characters\n\
                     listed below\n\
-  -E              disable interpretation of those sequences in STRINGs\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
       fputs (VERSION_OPTION_DESCRIPTION, stdout);
       fputs (_("\
 \n\
-Without -E, the following sequences are recognized and interpolated:\n\
+With -e, the following sequences are recognized and interpolated:\n\
 \n\
   \\NNN   the character whose ASCII code is NNN (octal)\n\
   \\\\     backslash\n\
