--- coreutils-5.2.1/src/test.c.orig	2004-07-14 05:59:14.000000000 -0400
+++ coreutils-5.2.1/src/test.c	2004-07-14 05:59:53.000000000 -0400
@@ -287,7 +287,7 @@
  * term ::=
  *	'-'('h'|'d'|'f'|'r'|'s'|'w'|'c'|'b'|'p'|'u'|'g'|'k') filename
  *	'-'('L'|'x') filename
- *	'-t' [ int ]
+ *	'-t' int
  *	'-'('z'|'n') string
  *	string
  *	string ('!='|'=') string
@@ -1037,7 +1037,7 @@
 "), stdout);
       fputs (_("\
   -S FILE     FILE exists and is a socket\n\
-  -t [FD]     file descriptor FD (stdout by default) is opened on a terminal\n\
+  -t FD       file descriptor FD is opened on a terminal\n\
   -u FILE     FILE exists and its set-user-ID bit is set\n\
   -w FILE     FILE exists and is writable\n\
   -x FILE     FILE exists and is executable\n\
