--- coreutils-4.5.3.orig/src/install.c
+++ coreutils-4.5.3/src/install.c
@@ -532,7 +532,11 @@
       error (EXIT_FAILURE, errno, _("fork system call failed"));
       break;
     case 0:			/* Child. */
+#ifdef __ELF__
+      execlp ("strip", "strip", "--remove-section=.comment", "--remove-section=.note", path, NULL);
+#else
       execlp ("strip", "strip", path, NULL);
+#endif
       error (EXIT_FAILURE, errno, _("cannot run strip"));
       break;
     default:			/* Parent. */
