--- coreutils-5.0/src/uname.c.orig	2003-05-13 11:39:38.000000000 -0400
+++ coreutils-5.2.1/src/uname.c	2003-05-13 11:40:15.000000000 -0400
@@ -91,8 +91,6 @@
   {"release", no_argument, NULL, 'r'},  /* Obsolescent.  */
   {"kernel-version", no_argument, NULL, 'v'},
   {"machine", no_argument, NULL, 'm'},
-  {"processor", no_argument, NULL, 'p'},
-  {"hardware-platform", no_argument, NULL, 'i'},
   {"operating-system", no_argument, NULL, 'o'},
   {GETOPT_HELP_OPTION_DECL},
   {GETOPT_VERSION_OPTION_DECL},
@@ -119,8 +117,6 @@
       fputs (_("\
   -v, --kernel-version     print the kernel version\n\
   -m, --machine            print the machine hardware name\n\
-  -p, --processor          print the processor type\n\
-  -i, --hardware-platform  print the hardware platform\n\
   -o, --operating-system   print the operating system\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
@@ -189,14 +185,6 @@
 	  toprint |= PRINT_MACHINE;
 	  break;
 
-	case 'p':
-	  toprint |= PRINT_PROCESSOR;
-	  break;
-
-	case 'i':
-	  toprint |= PRINT_HARDWARE_PLATFORM;
-	  break;
-
 	case 'o':
 	  toprint |= PRINT_OPERATING_SYSTEM;
 	  break;
@@ -237,53 +225,6 @@
 	print_element (name.machine);
     }
 
-  if (toprint & PRINT_PROCESSOR)
-    {
-      char const *element = unknown;
-#if HAVE_SYSINFO && defined SI_ARCHITECTURE
-      {
-	static char processor[257];
-	if (0 <= sysinfo (SI_ARCHITECTURE, processor, sizeof processor))
-	  element = processor;
-      }
-#endif
-#ifdef UNAME_PROCESSOR
-      if (element == unknown)
-	{
-	  static char processor[257];
-	  size_t s = sizeof processor;
-	  static int mib[] = { CTL_HW, UNAME_PROCESSOR };
-	  if (sysctl (mib, 2, processor, &s, 0, 0) >= 0)
-	    element = processor;
-	}
-#endif
-      print_element (element);
-    }
-
-  if (toprint & PRINT_HARDWARE_PLATFORM)
-    {
-      char const *element = unknown;
-#if HAVE_SYSINFO && defined SI_PLATFORM
-      {
-	static char hardware_platform[257];
-	if (0 <= sysinfo (SI_PLATFORM,
-			  hardware_platform, sizeof hardware_platform))
-	  element = hardware_platform;
-      }
-#endif
-#ifdef UNAME_HARDWARE_PLATFORM
-      if (element == unknown)
-	{
-	  static char hardware_platform[257];
-	  size_t s = sizeof hardware_platform;
-	  static int mib[] = { CTL_HW, UNAME_HARDWARE_PLATFORM };
-	  if (sysctl (mib, 2, hardware_platform, &s, 0, 0) >= 0)
-	    element = hardware_platform;
-	}
-#endif
-      print_element (element);
-    }
-
   if (toprint & PRINT_OPERATING_SYSTEM)
     print_element (HOST_OPERATING_SYSTEM);
 
